File: FBControlCore/Commands/FBProcessSpawnCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBProcessSpawnCommands.h"

#import "FBControlCoreError.h"
#import "FBControlCoreLogger.h"
#import "FBDataBuffer.h"
#import "FBProcess.h"
#import "FBProcessIO.h"
#import "FBProcessSpawnConfiguration.h"

@implementation FBProcessSpawnCommandHelpers

+ (dispatch_queue_t)queue
{
  return dispatch_queue_create("com.facebook.fbcontrolcore.process_spawn_helpers", DISPATCH_QUEUE_CONCURRENT);
}

+ (FBFuture<NSNumber *> *)launchAndNotifyOfCompletion:(FBProcessSpawnConfiguration *)configuration withCommands:(id<FBProcessSpawnCommands>)commands
{
  return [[commands
    launchProcess:configuration]
    onQueue:self.queue fmap:^(FBProcess *process) {
      return [process exitCode];
    }];
}

+ (FBFuture<NSString *> *)launchConsumingStdout:(FBProcessSpawnConfiguration *)configuration withCommands:(id<FBProcessSpawnCommands>)commands
{
  id<FBAccumulatingBuffer> consumer = FBDataBuffer.accumulatingBuffer;
  FBProcessIO *io = [[FBProcessIO alloc]
    initWithStdIn:configuration.io.stdIn
    stdOut:[FBProcessOutput outputForDataConsumer:consumer]
    stdErr:configuration.io.stdOut];
  FBProcessSpawnConfiguration *derived = [[FBProcessSpawnConfiguration alloc]
    initWithLaunchPath:configuration.launchPath
    arguments:configuration.arguments
    environment:configuration.environment
    io:io
    mode:configuration.mode];
  return [[self
    launchAndNotifyOfCompletion:derived withCommands:commands]
    onQueue:self.queue map:^(NSNumber *_) {
      return [[NSString alloc] initWithData:consumer.data encoding:NSUTF8StringEncoding];
    }];
}

+ (void)resolveProcessFinishedWithStatLoc:(int)statLoc inTeardownOfIOAttachment:(FBProcessIOAttachment *)attachment statLocFuture:(FBMutableFuture<NSNumber *> *)statLocFuture exitCodeFuture:(FBMutableFuture<NSNumber *> *)exitCodeFuture signalFuture:(FBMutableFuture<NSNumber *> *)signalFuture processIdentifier:(pid_t)processIdentifier configuration:(FBProcessSpawnConfiguration *)configuration queue:(dispatch_queue_t)queue logger:(id<FBControlCoreLogger>)logger
{
  [logger logFormat:@"Process %d (%@) has exited, tearing down IO...", processIdentifier, configuration.processName];
  [[attachment
    detach]
    onQueue:queue notifyOfCompletion:^(id _) {
      [logger logFormat:@"Teardown of IO for process %d (%@) has completed", processIdentifier, configuration.processName];
      [statLocFuture resolveWithResult:@(statLoc)];
      if (WIFSIGNALED(statLoc)) {
        int signalCode = WTERMSIG(statLoc);
        NSString *message = [NSString stringWithFormat:@"Process %d (%@) exited with signal %d", processIdentifier, configuration.processName, signalCode];
        [logger log:message];
        NSError *error = [[FBControlCoreError describe:message] build];
        [exitCodeFuture resolveWithError:error];
        [signalFuture resolveWithResult:@(signalCode)];
      } else {
        int exitCode = WEXITSTATUS(statLoc);
        NSString *message = [NSString stringWithFormat:@"Process %d (%@) exited with code %d", processIdentifier, configuration.processName, exitCode];
        [logger log:message];
        NSError *error = [[FBControlCoreError describe:message] build];
        [signalFuture resolveWithError:error];
        [exitCodeFuture resolveWithResult:@(exitCode)];
      }
    }];
}

@end

File: FBControlCore/Commands/FBProcessSpawnCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

/// Set as the error on whichever of `exitCode` / `signal` did not happen.
public enum FBProcessTerminationError: Error {
  case exitedWithSignal(processIdentifier: pid_t, processName: String, signal: Int32)
  case exitedWithCode(processIdentifier: pid_t, processName: String, exitCode: Int32)
}

extension FBProcessTerminationError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .exitedWithSignal(processIdentifier, processName, signal):
      return "Process \(processIdentifier) (\(processName)) exited with signal \(signal)"
    case let .exitedWithCode(processIdentifier, processName, exitCode):
      return "Process \(processIdentifier) (\(processName)) exited with code \(exitCode)"
    }
  }
}

@objc(FBProcessSpawnCommandHelpers)
public final class FBProcessSpawnCommandHelpers: NSObject {

  @objc
  public class func resolveProcessFinished(
    withStatLoc statLoc: Int32,
    inTeardownOfIOAttachment attachment: FBProcessIOAttachment,
    statLocFuture: FBMutableFuture<NSNumber>,
    exitCodeFuture: FBMutableFuture<NSNumber>,
    signalFuture: FBMutableFuture<NSNumber>,
    processIdentifier: pid_t,
    configuration: FBProcessSpawnConfiguration,
    queue: DispatchQueue,
    logger: (any FBControlCoreLogger)?
  ) {
    // One line per exit: the outcome below logs inside this completion handler,
    // so it inherently reports after IO teardown has finished. The separate
    // tearing-down/completed lines tripled every process exit for no
    // additional information.
    unsafeBitCast(attachment.detach(), to: FBFuture<AnyObject>.self)
      .onQueue(
        queue,
        notifyOfCompletion: { _ in
          statLocFuture.resolve(withResult: NSNumber(value: statLoc))
          let wstatus = statLoc & 0x7f // _WSTATUS
          if wstatus != 0x7f /* _WSTOPPED */ && wstatus != 0 {
            // WIFSIGNALED
            let signalCode = statLoc & 0x7f // WTERMSIG
            let error = FBProcessTerminationError.exitedWithSignal(processIdentifier: processIdentifier, processName: configuration.processName, signal: signalCode)
            logger?.log(error.localizedDescription)
            exitCodeFuture.resolveWithError(error)
            signalFuture.resolve(withResult: NSNumber(value: signalCode))
          } else {
            let exitCode = (statLoc >> 8) & 0xff // WEXITSTATUS
            let error = FBProcessTerminationError.exitedWithCode(processIdentifier: processIdentifier, processName: configuration.processName, exitCode: exitCode)
            logger?.log(error.localizedDescription)
            signalFuture.resolveWithError(error)
            exitCodeFuture.resolve(withResult: NSNumber(value: exitCode))
          }
        })
  }
}
